internal/operator/controller/gateway/controller.go

```diff
@@ -215,7 +211,7 @@ func (r *reconciler) ensureGateway(ctx context.Context, gw *gatewayv1alpha1.Gate
 	}
 
 	// configmap error/logging messages are different, hence not using handleResult
-	if err := objcm.Ensure(ctx, cli, objcm.NewCfgForGateway(gw)); err != nil {
+	if err := objcm.Ensure(ctx, cli, objcm.NewCfgForGateway(gw, operatorv1alpha1.GatewayClassControllerRef)); err != nil {
```

Comment:
Should the `NewCfgForGateway` use the passed in controllerName?

Reply:
i used this shortcut b/c the operator already expects a gw class with that controller name string (otherwise it doesnt reconcile it) and the operator checks that the gw class referenced by the gateway has that same string as well

I would say maybe as part of our next refactors to remove the gw controllers for this we should do the change to the Contour CRD Gateway ref field and rename/repurpose it to be a controller string (I'll add that to the proposal doc)

Reply:
It's just that if you have 2 instances running in a cluster, they'd end up with the same controllerName wouldn't they? That wouldn't allow Contour to process since it'd find duplicates.

Reply:
so with:
- gatewayclass name `gc1` (controller string `projectcontour.io/contour-operator`)
- gateway `ns1/gw1` referring to `gc1`
- gatewayclass name `gc2` (controller string `projectcontour.io/contour-operator`)
- gateway `ns2/gw2` referring to `gc2`
what happens in the operator:
- sees `gc1` and reconciles, its got the correct controller string (no longer sets status)
- sees `gc2` and reconciles, its got the correct controller string (no longer sets status)
- sees `gw1` and creates a `ns1/contour1` for it, passing `projectcontour.io/contour-operator` and `ns1/gw1`
- sees `gw2` and creates a `ns2/contour2` for it, passing `projectcontour.io/contour-operator` and `ns2gw2`
what happens in contour controller:
- `ns1/contour1` and `ns2/contour2` sees `gc1` and and both try to set status to admitted (this is weird/problematic)
- `ns1/contour1` and `ns2/contour2` sees `gc2` and also both try to set status to admitted? (we have't solved multiple gatewayclasses yet and this is weird/problematic)
- `ns1/contour1` sees `ns1/gw1` and reconciles
- `ns2/contour2` sees `ns2/gw2` and reconciles

Seems like some options are:
- operator doesn't hard code the controller string it looks for, should be configurable by user (repurpose or rename gateway ref like you said above)
  - Contour should also implement conflict handling of multiple gatewayclasses with the same controller string
- ditch the idea of using the controller string and just use gatewayclass names in Contour
- regardless of the above, the Operator should not allow users to create multiple contour instances referencing the *same* gatewayclass, by controller string or gatewayclass name (since I dont think that really makes sense)

Reply:
I don't know how different the scenario is now than what it was before, I think there were issues before with having multiple contours, def something to fix but we're helped out by the fact that contour does reconcile a gw class but just sets status, currently contour's gateway processing doesn't check the admitted status of the gw class a gw references

Reply:
Could we (for now) allow the operator to process any GatewayClass with a controller string starting with `projectcontour.io/`? And then if a user wants multiple Contours, they can use different controllerName strings starting with `projectcontour.io/` for each GatewayClass? Not sure if that would break things or not but maybe would work and allow >1 Contour to be created.

Reply:
so just testing things out, multiple contours works though it is weird

Reply:
Huh. It looks to me from the Contour code like the 2nd GatewayClass & Gateway created will take effect in *both* Contours because every time a new one comes in, the controllers `OnAdd` it to the cache/DAG, which overwrites whatever was already there - so if the 2nd GatewayClass/Gateway have identical specs to the first GatewayClass/Gateway, then yeah everything would appear to work, but otherwise the 1st Contour would end up with the 2nd Gateway's configuration too. Looking for an easy way to verify that..

Reply:
yeah, i may have to take back what i said about things working I had something broken locally, I think the multi-contour tests in the operator before weren't testing any requests actually work etc.

Reply:
yeah the multiple contour issue isn't going to be resolvable until the operator and Contour change significantly:
- operator no longer has a gw/gc controller
- operator doesn't allow users to create multiple Contours with the same controller string
- contour handles multiple gateways with the same gc name properly
- contour handles multiple gatewayclasses with the same controller string properly

Reply:
though checking the config dump, seems like the different contours are correctly also b/c the cache does filter on the named gateway that it is configured with

Reply:
ok added a test that verifies that 2 contours with different gateways can work still

Reply:
Ah was wondering about https://github.com/projectcontour/contour-operator/issues/411#issuecomment-885070813 but saw you've thought about that already

Reply:
The operator should make this string unique to the instance it's deploying. So maybe something like `projectcontour.io/<namespace>/<name>` of the instance? 

Contour expects this string to be unique per GatewayClass/Gateway.